x/marker/spec/12_transfers.md

```diff
@@ -0,0 +1,264 @@
+    transq{{"Does Sender have\ntransfer for Denom?"}}
+    mreqattr{{"Does Denom have\nrequired attributes?"}}
+    treqattr{{"Does Receiver have\nthose attributes?"}}
+    ok(["Denom transfer allowed."])
+    style ok fill:#bbffaa,stroke:#1b8500,stroke-width:3px
+    denied(["Send denied."])
+    style denied fill:#ffaaaa,stroke:#b30000,stroke-width:3px
+    transq -->|yes| ok
+    transq -.->|no| mreqattr
+    mreqattr -->|yes| treqattr
+    mreqattr -.->|no| denied
+    treqattr -->|yes| ok
+    treqattr -.->|no| denied
+
+    linkStyle 3,5 stroke:#b30000,color:#b30000
+    linkStyle 0,4 stroke:#1b8500,color:#1b8500
+```
+
+If the `Send` is allowed, and the `Receiver` is a quarantined account, the quarantine module's `SendRestrictionFn` will then change the `Send`'s destination to `QFH` (the Quarantined-funds-holder account) and make a record of the transfer. The `Send` then transfers funds from the `Sender` to `QFH`.
+
+The marker's `SendRestrictionFn` should never have `QFH` as a `Receiver`. The only way this would happen is if `MsgSend` is used to send funds directly to `QFH`.
+
+#### Accepting Quarantined Restricted Coins
+
+Once funds have been sent to `QFH`, the `Receiver` will probably want to accept them, and have them sent to their account. They issue an `Accept` to the quarantine module which utilizes the bank module's `Send` functionality to try to transfer funds from `QFH` to the `Receiver`.
+
+`QFH` is a bypass account. Since `Receiver` is a quarantined account, we can assume that it is neither a marker nor bypass account. So, the `validateSendDenom` flow can be simplified to this for restricted coins.
+
+```mermaid
+%%{ init: { 'flowchart': { 'curve': 'monotoneY'} } }%%
+flowchart LR
+    vsd[["validateSendDenom(Sender, Receiver, Denom)"]]
+    mreqattr{{"Does Denom have\nrequired attributes?"}}
+    treqattr{{"Does Receiver have\nthose attributes?"}}
+    ok(["Denom transfer allowed."])
+    style ok fill:#bbffaa,stroke:#1b8500,stroke-width:3px
+    denied(["Send denied."])
+    style denied fill:#ffaaaa,stroke:#b30000,stroke-width:3px
+    mreqattr -->|yes| treqattr
+    mreqattr -.->|no| ok
+    treqattr -->|yes| ok
+    treqattr -.->|no| denied
+
+    linkStyle 3 stroke:#b30000,color:#b30000
+    linkStyle 1,2 stroke:#1b8500,color:#1b8500
+
+```
+
+If the `Send` is allowed, the requested funds are transferred from `QFH` to `Receiver`.
+
+If the `Send` is denied, the funds remain with `QFH`.
+
+An important subtle part of this process is the rechecking of `Receiver` attributes. It's possible for the initial send to be okay (causing funds to be quarantined), then later, during this `Accept`, the send is not okay, and the quarantined funds are effectively locked with`QFH` until the `Receiver` gets the required attributes.
+
+If the marker does not have required attributes though, it's assumed that they were originally sent by someone with transfer authority, so they are allowed to continue to from here too.
```

Comment:
I fixed the first paragraph, but left off the note.

Removing funds from the `QFH` account without updating quarantine records will break an invariant.

`authz` cannot be used like that to move funds out of the `QFH`. The only way funds can leave the `QFH` is through the quarantine `Accept` endpoint. Even if `authz` is used to allow a third party to call `Accept`, the marker restrictions only know about the source account (`QFH`) and the destination (the `toAddr`).

Funds can become stuck in the `QFH` if there are required attributes but the `toAddr` doesn't have them. The only workarounds would be to give the `toAddr` the required attributes or grant `QFH` transfer permission. The latter is probably a very bad idea.

If these stuck funds become a problem, we could either add some functionality to `MsgTransferRequest` or create a new `Rescue` endpoint somewhere. Either way, quarantine records need to be updated as when the funds are moved out of `QFH`.

Reply:
This brings up a good point… at the moment the authz is for the msgtransfer endpoint, but we also have attribute checks on acceptance and potentially disabled ability on source to send without a corresponding ability to express these rights in authz.  That ability would be needed here for the above as the intent wasn’t to move the funds from the quarantine unilaterally but instead to satisfy the recipient eligibility criteria.

Reply:
The required attributes and deny-list checks are specific to the account receiving funds and account the funds are coming out of (respectively). The former can be bypassed using the transfer permission. The latter is only bypassable by using the `MsgTransferRequest` endpoint, though I wonder if even that shouldn't be allowed.

If an account does not have the required attributes, but funds should still be sent to it, there's a few ways to do it:
1. The current holder of the funds is given transfer permission. They can then use any method of sending the funds that they want.
2. The current holder of the funds is conveyed transfer permission through authz using a `GenericAuthorization` on the `MsgTransferRequest` messages. They then submit an authz `Exec` with a `MsgTransferRequest` identifying the granter as the `admin`.
3. The current holder of the funds issues a `MarkerTransferAuthorization` to someone with transfer permission who then sends a `MsgTransferRequest` on their behalf.

The deny list check isn't done during a `MsgTransferRequest`. It's only applied when the `Send` is initiated by other means (e.g. `MsgSend`). That check is done before checking for transfer authority though, so if someone is on the deny list, they only way the funds can move out of their account is through a `MsgTransferRequest`. So option 2 or 3 above would work there too.

All authz does is allow a message to have a different signer from the one contained in the message. Usually, by the time we're processing a message, that checking as already been done and we don't know the actual signer, just the "signer" identified in the msg. During a quarantine `Accept`, the signer is the `toAddr`. However, for an `Accept` to be processed, there must be a record of the quarantined funds to that address, and the funds can only go to that address. There's no way for it to know of another address that might have tried to convey transfer permission in order to allow the `Accept` to go through. During a `MsgSend`, the same is true. There's no way for a third address to be involved conveying transfer permission.

Currently, the only way to get a third address involved like that is through `MsgTransferRequest`. But that can't be used for quarantined funds because `QFH` is a module account (plus, doing so would break an invariant). If we don't mind the marker module having a dependency on the quarantine module, we can add a check in `MsgTransferRequest` to identify when the `from` is `QFH` and properly tell the quarantine module about the fund movement. Otherwise, somewhere, we'd have to create a new `Rescue` endpoint that would provide the ability to get stuck funds out of quarantine, applying the appropriate restrictions and properly updating the quarantine module.

---

Should we be doing the deny-list check during a `MsgTransferRequest`? I feel like we should since transfer permission does not bypass that check during other sends. Or maybe transfer permission should bypass the deny-list check during other sends too. But I feel like the spirit of the deny-list is the same as a hold; i.e. those funds cannot leave the account until they're off the deny list.

Reply:
No, we should not be restricting `MsgTransferRequest` as it is more correctly `MsgTransfer[Authority]Request` meaning you must be a holder of the `transfer` grant (directly or indirectly) and you have complete authority to perform transfers.  Literally the Transfer Authority sets the rules and can make exceptions to them.

Reply:
Summarizing an outside conversation in here for the record.

Let's say `chickens` are a restricted coin, and I have a bunch of them. I have transfer permission for them, but am also on the deny list for them.
* If I `MsgSend` from: me, to: you, amount: `5chickens`, it will fail because I'm on the deny list.
* If I `MsgTransferRequest` from: me, to: you, admin: me, amount: 5chickens, it will go through because it doesn't check the deny list.

This is expected behavior because the `MsgTransferRequest` is sort of a `sudo bank send`. Someone with transfer but on the deny list needs to recognize that they're possibly bypassing some restrictions to do their transfer.

Reply:
As for the issue with funds getting stuck in quarantine, I feel like that would be better handled with a separate issue and PR since there will be some design and implementation decisions to make.

Reply:
I created #1654 for the stuck-in-quarantine thing.